src/Physics/QuasiElastic/EventGen/QELEventGeneratorSM.cxx

```diff
@@ -116,28 +117,29 @@ void QELEventGeneratorSM::ProcessEventRecord(GHepRecord * evrec) const
   fXSecModel = evg->CrossSectionAlg();
 
   // heavy nucleus is nucleus that heavier than hydrogen and deuterium
-  bool isHeavyNucleus = tgt->A()>=3;
+  bool isHeavyNucleus = tgt->A()>3;
 
   sm_utils->SetInteraction(interaction);
   // phase space for heavy nucleus is different from light one
-  fkps = isHeavyNucleus?kPSQ2vfE:kPSQ2fE;
+  fkps = isHeavyNucleus?kPSQ2vpfE:kPSQ2fE;
   Range1D_t rQ2 = sm_utils->Q2QES_SM_lim();
   // Try to calculate the maximum cross-section in kinematical limits
   // if not pre-computed already
-  double xsec_max1  = (fGenerateUniformly) ? -1 : this->MaxXSec(evrec);
-  double xsec_max2  = (fGenerateUniformly) ? -1 : (rQ2.max<fQ2Min)? 0: this->MaxXSec2(evrec);// this make correct calculation of probability
-  double vmax= isHeavyNucleus?this->MaxDiffv(evrec) : 0.;
+  double xsec_max1  = fGenerateUniformly ? -1 : this->MaxXSec(evrec);
+  // this make correct calculation of probability
+  double xsec_max2  = fGenerateUniformly ? -1 : (rQ2.max<fQ2Min)? 0:this->MaxXSec(evrec, 1);
+  double dvmax= isHeavyNucleus ? this->MaxXSec(evrec, 2) : 0.;
 
 
-  // generate Q2, v
-  double gQ2, v, xsec;
+  // generate Q2, v, pF
+  double Q2, v, kF, xsec;
   unsigned int iter = 0;
   bool accept = false;
   TLorentzVector q;
   while(1)
   {
      LOG("QELEvent", pINFO) << "Attempt #: " << iter;
-     if(iter > kRjMaxIterations)
+     if(iter > 100*kRjMaxIterations)
```

Comment:
The constant `kRjMaxIterations` is defined in `genie::controls` and set equal to 1000. In my opinion, it can be changed and set equal to 100,000. The only problem I see, this is when an `EventGenerator` fails to generate events at some energy, this setting slow down the speed of events generation. However, this extreme case would be meant that `EventGenerator` is coded with mistakes.

Reply:
Fixed